File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(tree, leftBound, rightBound):
            if not tree:
                return True
            
            if not (tree.val < rightBound and tree.val > leftBound):
                return False
            
            
            return dfs(tree.left, leftBound, tree.val) and dfs(tree.right, tree.val, rightBound)
        
        return dfs(root, float('-inf'), float('inf'))
        

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


bool isValidBST(struct TreeNode* root){
        
    bool dfs(struct TreeNode* tree, int left, int right, bool l_flag, bool r_flag) {
        if(!tree) {
            return true;
        }
        
        if (!((tree->val > left || l_flag) && (tree->val < right || r_flag))) {
            return false;
        }
        
        return dfs(tree->left, left, tree->val, l_flag, false) && dfs(tree->right, tree->val, right, false, r_flag);
    }
    
    return dfs(root, 0, 0, true, true);

}